<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Этап-1.-Обзор-данных" data-toc-modified-id="Этап-1.-Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Этап 1. Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Шаг-1.-Знакомство-с-данными" data-toc-modified-id="Шаг-1.-Знакомство-с-данными-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Шаг 1. Знакомство с данными</a></span></li></ul></li><li><span><a href="#Этап-2.-Предобработка-данных" data-toc-modified-id="Этап-2.-Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Этап 2. Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления" data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных" data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов" data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма" data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода" data-toc-modified-id="Шаг-2.6.-Категоризация-дохода-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита" data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита</a></span></li></ul></li><li><span><a href="#Этап-3.-Проверка-гипотез" data-toc-modified-id="Этап-3.-Проверка-гипотез-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Этап 3. Проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Проверка-гипотезы-№1" data-toc-modified-id="Проверка-гипотезы-№1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Проверка гипотезы №1</a></span></li><li><span><a href="#Проверка-гипотезы-№2" data-toc-modified-id="Проверка-гипотезы-№2-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка гипотезы №2</a></span></li><li><span><a href="#Проверка-гипотезы-№3" data-toc-modified-id="Проверка-гипотезы-№3-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Проверка гипотезы №3</a></span></li><li><span><a href="#Проверка-гипотезы-№4" data-toc-modified-id="Проверка-гипотезы-№4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Проверка гипотезы №4</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надежности заемщиков

**Цель исследования** — проверить четыре гипотезы:

1. Количество детей клиента влияет на факт погашения кредита в срок.
2. Семейное положение клиента влияет на факт погашения кредита в срок.
3. Уровень дохода клиента влияет на факт погашения кредита в срок.
4. Цель кредита влияет на факт погашения кредита в срок.

**Ход исследования**

Будем использовать входные данные от банка — статистика о платёжеспособности клиентов из файла `/datasets/data.csv`.
Перед проверкой гипотез нам понадобится обзор данных. 

Этапы исследования:
 1. Обзор данных.
 2. Предобработка данных.
 3. Проверка гипотез.

<div style="border:solid black 1px; padding: 20px">
    
**Описание данных**
    
- children — количество детей в семье 
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента 
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита
    
</div>    

## Этап 1. Обзор данных
### Шаг 1. Знакомство с данными

In [1]:
# импорт библиотеки pandas
import pandas as pd

# чтение файла с данными и сохранение в df
df = pd.read_csv('/datasets/data.csv')

# получение общей информации о данных в df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-block alert-info">
<font color='Black'>

Общее количество строк в датафрейме 21525, столбцов 12, встречающиеся типы данных: вещественные и целые числа, а также строковые значения. 
    
В столбцах `days_employed` и `total_income` имеются пропущенные значения, так как в данных столбцах только 19351 непропущенных значений, что меньше общего количества строк в датафрейме.

In [2]:
# получение первых 10 строк датафрейма df
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<div class="alert alert-block alert-info">
<font color='Black'>

В ходе визуального анализа первых 10 строк обнаружили отрицательные значения в столбце `days_employed`, по своей сути значения в данном столбце должны принимать неотрицательные значения. 
Определим, как часто в данном столбце встречаются отрицательные значения

In [3]:
# выберем в df строки с отрицательными значением дней общего трудового стажа с помощью логической индексации
# посчитаем количество таких строк методом count() по столбцу days_employed
df[df['days_employed'] < 0]['days_employed'].count()

15906

<a id='survey'></a>
<div class="alert alert-block alert-info">
<font color='Black'>
    
Получили 15906 строк с отрицательным значением общего трудового стажа, это составляет около 74% всех строк, что очень много.
Попробуем обработать данные значения на этапе предобработки данных.

<div style="border:solid green 3px; padding: 20px">

Провели обзор данных: общее количество строк в датафрейме 21525, столбцов 12, наименования столбцов отражают суть, содержащихся в них данных, и имеют корректные названия, встречающиеся типы данных - вещественные и целые числа, а также строковые значения, типы данных столбцов по своей сути соответствуют их содержанию. 
    
Выявили следующие проблемы:
- пропущенные значения в двух столбцах `days_employed` и `total_income`, по 2174 пропущенных значения в каждом из столбцов,
- отрицательные значения в столбце `days_employed` в количестве 15906, что составляет около 74% всех строк.

## Этап 2. Предобработка данных
### Шаг 2.1 Заполнение пропусков

<div class="alert alert-block alert-info">
<font color='Black'>

На этапе обзора данных в столбцах `days_employed` и `total_income` были обнаружены пропущенные значения, так как в данных столбцах только 19351 непропущенных значений, что меньше общего количества строк в датасете 21525.

Перед заполнением пропусков проанализируем, какие значения имеются в данных столбцах. 

На этапе [обзора](#survey) данных мы увидели, что в столбце `days_employed` имеется довольно большое количество отрицательных значений. 

Стоит отметить, что данный столбец не важен для проверки гипотез исследования, но попробуем проанализировать значения в данном столбце.

Можно предположить, что минус в отрицательных значениях мог возникнуть при заполнении базы данных о клиентах, например, основной источник с данными о стаже содержал дефисы, при переносе данных в базу дефисы превратились в минусы, так как тип данного столбца числовой.

In [4]:
# найдем максимальное значение в столбце days_employed
df['days_employed'].max()

401755.40047533

<div class="alert alert-block alert-info">
<font color='Black'>

Для трудового стажа в днях 401755 это очень большое значение, что в годах составляет примерно 1100 лет! Вероятно, снова ошибка, возникшая в результате переноса данных в базу данных банка, возможно это другие единицы измерения, например, часы.

Попробуем оценить количество строк с аномально высокими значениями в столбце `days_employed`.

In [5]:
#предположим, что максимальный стаж составляет примерно 80 лет или 29200 дней
#посчитаем количество строк в столбце days_employed со значением больше, чем предполагаемый максимальный стаж
df[df['days_employed'] > 29200]['days_employed'].count()

3445

<div class="alert alert-block alert-info">
<font color='Black'>

Получаем, что в столбце `days_employed` 3445 аномально высоких значений и 15906 отрицательных значений, что в сумме составляет общее число непропущенных значений в данном столбце 19351. 

In [6]:
# получение первых 10 строк датафрейма df с аномально высокми значениями стажа
df[df['days_employed'] > 29200].head(10)

children  days_employed  dob_years education  education_id  \
4          0  340266.072047         53   среднее             1   
18         0  400281.136913         53   среднее             1   
24         1  338551.952911         57   среднее             1   
25         0  363548.489348         67   среднее             1   
30         1  335581.668515         62   среднее             1   
35         0  394021.072184         68   среднее             1   
50         0  353731.432338         63   среднее             1   
56         0  370145.087237         64   среднее             1   
71         0  338113.529892         62   среднее             1   
78         0  359722.945074         61    высшее             0   

            family_status  family_status_id gender income_type  debt  \
4        гражданский брак                 1      F   пенсионер     0   
18         вдовец / вдова                 2      F   пенсионер     0   
24  Не женат / не замужем                 4      F   пенсионер     0   
25        женат / замужем                 0      M   пенсионер     0   
30        женат / замужем                 0      F   пенсионер     0   
35       гражданский брак                 1      M   пенсионер     0   
50        женат / замужем                 0      F   пенсионер     0   
56         вдовец / вдова                 2      F   пенсионер     0   
71        женат / замужем                 0      F   пенсионер     0   
78        женат / замужем                 0      M   пенсионер     0   

     total_income                                purpose  
4   158616.077870                        сыграть свадьбу  
18   56823.777243     на покупку подержанного автомобиля  
24  290547.235997  операции с коммерческой недвижимостью  
25   55112.757732                   покупка недвижимости  
30  171456.067993  операции с коммерческой недвижимостью  
35   77805.677436                  на проведение свадьбы  
50   92342.730612                             автомобили  
56  149141.043533                            образование  
71   43929.696397                             автомобили  
78  175127.646000                   сделка с автомобилем

<div class="alert alert-block alert-info">
<font color='Black'>

Видим, что в первых 10 строках датафрейма с аномально высокими значениями стажа в столбце `income_type` находится значение "пенсионер". Сгруппируем датафрейм с аномально высокми значениями стажа по столбцу `income_type` и посчитаем количество строк в каждой группе, аналогично поступим с датафреймом с отрицательными значениями общего трудового стажа.

In [7]:
# cгруппируем строки датафрейма с аномально высокими значениями стажа по столбцу income_type 
# посчитаем количество строк по кажой группе методом count()
df[df['days_employed'] > 29200].groupby('income_type').count()

children  days_employed  dob_years  education  education_id  \
income_type                                                                
безработный         2              2          2          2             2   
пенсионер        3443           3443       3443       3443          3443   

             family_status  family_status_id  gender  debt  total_income  \
income_type                                                                
безработный              2                 2       2     2             2   
пенсионер             3443              3443    3443  3443          3443   

             purpose  
income_type           
безработный        2  
пенсионер       3443

In [8]:
# cгруппируем строки датасфрейма с отрицательными значениями стажа по столбцу income_type 
# посчитаем количество строк по кажой группе методом count()
df[df['days_employed'] < 0].groupby('income_type').count()

children  days_employed  dob_years  education  education_id  \
income_type                                                                    
в декрете               1              1          1          1             1   
госслужащий          1312           1312       1312       1312          1312   
компаньон            4577           4577       4577       4577          4577   
предприниматель         1              1          1          1             1   
сотрудник           10014          10014      10014      10014         10014   
студент                 1              1          1          1             1   

                 family_status  family_status_id  gender   debt  total_income  \
income_type                                                                     
в декрете                    1                 1       1      1             1   
госслужащий               1312              1312    1312   1312          1312   
компаньон                 4577              4577    4577   4577          4577   
предприниматель              1                 1       1      1             1   
сотрудник                10014             10014   10014  10014         10014   
студент                      1                 1       1      1             1   

                 purpose  
income_type               
в декрете              1  
госслужащий         1312  
компаньон           4577  
предприниматель        1  
сотрудник          10014  
студент                1

<div class="alert alert-block alert-info">
<font color='Black'>
    
Группировки показали, что аномально высокие значения стажа имеются по пенсионерам и безработным, а отрицательные значения стажа по всем остальным типам занятности. 

Изменим отрицательные значения трудового стажа на положительные, используя функцию модуля числа.

Аномально высокие значения стажа оставим без изменений, можно действовать из предположения, что данные значения представлены в часах и перевести их в дни, но лучше подробности выяснить у заказчика исследования. 
Результаты текущего исследования не зависят от значений столбца `days_employed`.

In [9]:
# применим к числовому столбцу days_employed функцию, возвращающую модуль числа
df['days_employed'] = abs(df['days_employed'])

# проверим минимальное значение в столбце days_employed, оно должно быть положительным
df['days_employed'].min()

24.14163324048118

In [10]:
# проверим минимальное значение в столбце total_income, оно должно быть положительным
df['total_income'].min()

20667.26379327158

<div class="alert alert-block alert-info">
<font color='Black'>

Минимальное значение в столбце `total_income` положительное, делаем вывод, что значения в данном столбце соответствуют действительности.
    
Проверим, какую долу составляют пропущенные значения в столбцах `days_employed` и `total_income`.

In [11]:
# выберем строки датафрейма с пропущенными значениями в столбце 'days_employed'
# вычислим их количество с помощью функции len 
# разделим полученное количество строк с пропущенными значениями на общее количество строк
len(df[df['days_employed'].isna()]) / df.shape[0]

0.10099883855981417

In [12]:
# найдем количество строк, в которых пропущены значения в столбцах `days_employed` и `total_income` одновременно
len(df[(df['days_employed'].isna()) & (df['total_income'].isna())]) 

2174

<div class="alert alert-block alert-info">
<font color='Black'>
    
Доля пропущенных значений в каждом из столбцов `days_employed` и `total_income` составляет около 10%, мы видим, что пропущенные значения в данных столбцах встречаются в одних и тех же строках (2174 шт), удаление 10% строк может повлиять на результаты исследования.
Возможно пропуски возникли на этапе заполнения данных в системе банка (человеческий фактор), либо на этапе выгрузки данных в файл (технологический фактор).

Пропущенные значения в количественных переменных принято заполнять значениями, характеризующими состояние выборки, для того чтобы поведение данных в выборке осталось неизменным. Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства, поэтому будем использовать медиану - значение, которое делит выборку пополам.


In [13]:
# заполним пропущенные значения медианным значением соответствующего столбца с помощью метода fillna()
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())
df['total_income'] = df['total_income'].fillna(df['total_income'].median())

In [14]:
# проверим, что пропуски заполнены
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div style="border:solid green 3px; padding: 20px">
    
Обнаружили пропущенные значения в столбцах `days_employed` и `total_income`, пропущенные значения в данных столбцах встречаются в одних и тех же строках (2174 шт), доля пропущенных значений составляет около 10%, удаление 10% строк может повлиять на результаты исследования, поэтом данные пропуски были обработаны.
Возможно, пропуски возникли на этапе заполнения данных в системе банка (человеческий фактор), либо на этапе выгрузки данных в файл (технологический фактор).

Перед заполнением пропусков проверили значения в столбцах `days_employed` и `total_income`, выявили отрицательные значения, а также  аномально высокие значения в `days_employed`. Отрицательные значения заменили абсолютным значением, аномально высокие значения оставили без изменений, так как значения столбца `days_employed` не влияют на результаты исследования.
    
Пропущенные значения заполнили медианным значением по соответствующему столбцу. Для того чтобы поведение данных в выборке осталось неизменным, выбрали метод заполнения пропусков характерным значением для количественных переменных. Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства, поэтому использовали медиану - значение, которое делит выборку пополам.

### Шаг 2.2 Проверка данных на аномалии и исправления

<div class="alert alert-block alert-info">
<font color='Black'>

Проверим данные в столбцах с категориальными переменными на уникальность значений и на частоту повторения данных значений.
Затем проверим данных в столбцах с количественными переменными, например, узнаем минимальное и максимальное значения. 
Значения столбцов `days_employed` и `total_income` с количественными переменными были проверены на предыдущем шаге.

In [15]:
# проверка на уникальность значений в столбце education с категориальными переменными и на частоту повторения данных значений
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>
    
<a id='detect_duplicates'></a>
`education` — уровень образования клиента - имеются дубликаты, для их устранения необходимо привести значения столбца к нижнему регистру, удалим дубликаты [позже](#delete_duplicates).

In [16]:
# проверка на уникальность значений в столбце education_id с категориальными переменными и на частоту повторения данных значений
df['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

`education_id` — идентификатор уровня образования - значения корректны, соответствуют количеству уникальных значений в столбце `education`

In [17]:
# проверка на уникальность значений в столбце family_status с категориальными переменными и на частоту повторения данных значений
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

`family_status` — семейное положение - значения корректны, формат не очень удобен для последующей обработки, но имеется столбец `family_status_id`

In [18]:
# проверка на уникальность значений в столбце family_status_id с категориальными переменными и на частоту повторения данных значений
df['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

`family_status_id` — идентификатор семейного положения - значения корректны, соответствуют количеству уникальных значений в столбце `family_status`

In [19]:
# проверка на уникальность значений в столбце gender с категориальными переменными и на частоту повторения данных значений
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

`gender` — пол клиента - значение `XNA` в одной строке, скорее всего, пропущено, значения данного столбца не повлияют на результаты исследования, поэтом оставим его без имзменений.

In [20]:
# проверка на уникальность значений в столбце income_type с категориальными переменными и на частоту повторения данных значений
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

`income_type` — тип занятости - значения корректны

In [21]:
# проверка на уникальность значений в столбце debt с категориальными переменными и на частоту повторения данных значений
df['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

`debt` — имел ли задолженность по возврату кредитов - значения корректны, просчроченная задолженность либо имела место, либо отсутствовала. 

In [22]:
# проверка на уникальность значений в столбце purpose с категориальными переменными и на частоту повторения данных значений
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

<div class="alert alert-block alert-info">
<font color='Black'>
    
`purpose` — цель получения кредита - видим неявные дубликаты, например, `свадьба`, `на проведение свадьбы`, `сыграть свадьбу`, нужно унифицировать значения целей (в идеале при вводе данных цель должна выбираться из справочника)

In [23]:
# проверка на уникальность значений в столбце children с количественными переменными и на частоту повторения данных значений
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

`children` — количество детей в семье - имеются отрицательное значение `-1` по 47 строкам, сомнительное значение `20` по 76 строкам, вероятно это ошибки, возникшие на этапе ввода данных, например, в базу данных банка, либо при выгрузке датасета в файл.

Данный столбец влияет на результаты исследования, поэтому некорректные значения нужно обработать.

Предположим, что минус в значениях столбца это дефис, который возник на этапе импорта данных в базу данных банка, заменим в столбце `children` значение `-1` на `1`. 

Поскольку количество строк с аномальным значением `20` меньше 1%, то заменим в столбце `children` значение `20` на медианное значение в данном столбце.

In [24]:
# заменим в столбце children значение -1 на 1
# заменим в столбце children значение 20 на медианное значение
df['children'].replace(-1, 1, inplace=True)
df['children'].replace(20, df['children'].median(), inplace=True)

In [25]:
# проверка на уникальность значения в столбце children еще раз
df['children'].value_counts()

0    14225
1     4865
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

Теперь столбец `children` не содержит аномалий.

In [26]:
# проверка значений в столбце dob_years с количественными переменными
# поиск максимального значения
df['dob_years'].max()

75

In [27]:
# проверка значений в столбце dob_years с количественными переменными
# поиск минимального значения
df['dob_years'].min()

0

In [28]:
# подсчет количества строк с dob_years = 0
df[df['dob_years'] == 0]['dob_years'].count()

101

<div class="alert alert-block alert-info">
<font color='Black'>

`dob_years` — возраст клиента в годах - имеются отсутствующие значения в виде 0, вероятно это ошибки, возникшие на этапе ввода данных, например, в базу данных банка, либо при выгрузке датасета в файл.

Данный столбец не влияет на результаты исследования, поэтому оставим его без изменений.

<div style="border:solid green 3px; padding: 20px">
    
Выявили следующие аномалии или некорректные значения, которые скорректировали на данном шаге либо скорректируем на последующих шагах:    
    
- `education` — уровень образования клиента - имеются дубликаты, для их устранения необходимо привести значения столбца к нижнему регистру, удалим дубликаты [позже](#delete_duplicates);
- `purpose` — цель получения кредита - имеются неявные дубликаты, например, `свадьба`, `на проведение свадьбы`, `сыграть свадьбу`, нужно унифицировать значения целей (в идеале при вводе данных цель должна выбираться из справочника);
- `children` — количество детей в семье - обнаружили отрицательное значение `-1` по 47 строкам, сомнительное значение `20` по 76 строкам, вероятно это ошибки, возникшие на этапе ввода данных, например, в базу данных банка, либо при выгрузке датасета в файл. Данный столбец влияет на результаты исследования, поэтому некорректные значения были обработаны (заменили в столбце `children` значение `-1` на `1`, значение `20` на медианное значение данного столбца).
    
Следующие некорректные значения не корректировались, так как данные значения не влияют на результаты исследования:
    
- `gender` — пол клиента - значение `XNA` в одной строке, скорее всего, пропущено.
- `dob_years` — возраст клиента в годах - имеются отсутсвующие значения в виде 0, вероятно это ошибки, возникшие на этапе ввода данных, например, в базу данных банка, либо при выгрузке данных в файл.


### Шаг 2.3. Изменение типов данных

<div class="alert alert-block alert-info">
<font color='Black'>

Для удобства обработки данных заменим вещественный тип данных в столбце `total_income` на целочисленный.

In [29]:
# перевод значений столбца в целочисленный тип методом astype()
df['total_income'] = df['total_income'].astype('int')

In [30]:
# получение общей информации о данных в df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


<div style="border:solid green 3px; padding: 20px">
    
Тип данных столбца `income_type` изменили с вещественного на целочисленный для удобства обработки.

### Шаг 2.4. Удаление дубликатов

<div class="alert alert-block alert-info">
<font color='Black'>

Проверим датафрейм на наличие явных дубликатов.

In [31]:
# применим метод duplicated() и просуммируем число полученных дубликатов
df.duplicated().sum()

54

<div class="alert alert-block alert-info">
<font color='Black'> 
Обнаружили 54 явных дубликата, удалим дублирующие строки.

In [32]:
# удалим явные дубликаты с помощью drop_duplicates() и обновим индексы с помощью reset_index()
# и перезапишем датафрейм
df = df.drop_duplicates().reset_index(drop=True)

In [33]:
# применим метод duplicated() и просуммируем число полученных дубликатов
df.duplicated().sum()

0

<div class="alert alert-block alert-info">
<font color='Black'>
    
<a id='delete_duplicates'></a>
Теперь явные дубликаты в датасете отсутствуют. Появление явных дубликатов в датасете вполне логично, так как некоторые клиенты могут совпадать по набору имеющихся признаков. Возможно имеет смысл добавить поле с идентификатором клиента.

На [шаге 2.2](#detect_duplicates) были обнаружены неявные дубликаты в столбце `education`, для их устранения необходимо привести значения столбца к нижнему регистру.

In [34]:
# приведем все символы в столдце education к нижнему регистру
df['education'] = df['education'].str.lower()

# проверка на уникальность значений в столбце education с категориальными переменными и на частоту повторения данных значений
df['education'].value_counts()

среднее                15188
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

<div class="alert alert-block alert-info">
<font color='Black'>

После приведения данных в столбце `education` к нижнему регистру, неявные дубликаты в данном столбце отсутствуют. Причина появления неявных дубликатов, вероятно, человеческий фактор и ошибка возникла на этапе ввода данных в базу данных банка.

<div style="border:solid green 3px; padding: 20px">
    
Удалили из датафрейма явные дубликаты, а также избавились от неявных дубликатов в столбце `education` путем приведения соответствующих значений к нижнему регистру.    

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма

<div class="alert alert-block alert-info">
<font color='Black'>

Для столбцов `education` и `family_status` создадим так называемые "словари" для того, чтобы в основном датафрейме хранить только идентификаторы типов образования и семейного статуса.

In [35]:
# создадим новый датафрейм со столбцами education_id и education
education_dict = df [['education_id', 'education']]

# удалим дубликаты из вновь созданного датафрейма и обновим индексы
education_dict = education_dict.drop_duplicates().reset_index(drop=True)

# создадим новый датафрейм со столбцами family_status_id и family_status
family_status_dict = df [['family_status_id','family_status']]

# удалим дубликаты из вновь созданного датафрейма и обновим индексы
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)

In [36]:
# печать на экран "словаря" типов образования
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [37]:
# печать на экран "словаря" типов семейного статуса
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

In [38]:
# перезапишем датафрейм, уберем столбцы education и family_status, для которых создали "словари"
df = df[['children', 'days_employed', 'dob_years', 'education_id', 'family_status_id', 'gender', 
         'income_type', 'debt','total_income', 'purpose']]

In [39]:
# получение общей информации о данных в df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21471 entries, 0 to 21470
Data columns (total 10 columns):
children            21471 non-null int64
days_employed       21471 non-null float64
dob_years           21471 non-null int64
education_id        21471 non-null int64
family_status_id    21471 non-null int64
gender              21471 non-null object
income_type         21471 non-null object
debt                21471 non-null int64
total_income        21471 non-null int64
purpose             21471 non-null object
dtypes: float64(1), int64(6), object(3)
memory usage: 1.6+ MB


<div style="border:solid green 3px; padding: 20px">

Создали "словари" для типов образования и семейного статуса (`education`, `family_status`), в основном датафрейме оставили только идентификаторы типов.

### Шаг 2.6. Категоризация дохода

<div class="alert alert-block alert-info">
<font color='Black'>

На основании диапазонов, указанных ниже, создадим столбец `total_income_category` с категориями:
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

In [40]:
# определим функцию total_income_group, которая по значению дохода income возвращает соответствующую категорию
def total_income_group(income):
    if income <= 30000:
        return 'E'
    elif income <= 50000:
        return 'D'
    elif income <= 200000:
        return 'C'
    elif income <= 1000000:
        return 'B'
    else:
        return 'A'

In [41]:
# создадим новый столбец в df, применив функцию total_income_group к столбцу total_income
df['total_income_category'] = df['total_income'].apply(total_income_group)

In [42]:
# получение первых 10 строк датасета df
df.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

<a id='income'></a>
<div style="border:solid green 3px; padding: 20px">
    
В датафрейм добавили новый столбец, отражающий категорию дохода клиента `total_income_category`:
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

### Шаг 2.7. Категоризация целей кредита

<div class="alert alert-block alert-info">
<font color='Black'>

Значения в столбце `purpose` довольно разнообразны, что усложняет анализ. Но по сути можно выделить четыре категории целей. Для этого создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:
- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

In [43]:
# получим уникальные значения столбца purpose и их количество
df['purpose'].value_counts()

свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

In [44]:
# определим функцию purpose_group, которая в зависимости от цели purpose будет возвращать значение из 
# purpose_list = ['операции с автомобилем','операции с недвижимостью','проведение свадьбы','получение образования']
# при условии вхождения определенной подстроки в строку purpose
def purpose_group(purpose):
    purpose_list = ['операции с автомобилем',
                    'операции с недвижимостью',
                    'проведение свадьбы',
                    'получение образования']
    if 'автомобил' in purpose:
        return purpose_list[0]
    elif ('недвижимост' in purpose) or ('жиль' in purpose):
        return purpose_list[1]        
    elif 'свадьб' in purpose:
        return purpose_list[2]
    elif 'образовани' in purpose:
        return purpose_list[3]  
    else:
        return 'цель не определена'

In [45]:
# создадим новый столбец в df, применив функцию purpose_group к столбцу purpose
df['purpose_category'] = df['purpose'].apply(purpose_group)

In [46]:
# получим уникальные значения нового столбца purpose_category и их количество
df['purpose_category'].value_counts()

операции с недвижимостью    10814
операции с автомобилем       4308
получение образования        4014
проведение свадьбы           2335
Name: purpose_category, dtype: int64

<div style="border:solid green 3px; padding: 20px">
    
Добавили в датафрейм новый столбец `purpose_category`, который вклдючает в себя следующие категории:
- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.    

## Этап 3. Проверка гипотез

<div class="alert alert-block alert-info">
<font color='Black'>
 
Реализуем дополнительные функции, которые понадобятся для вывода результатов.

In [47]:
# определим функцию retun_rate, которая на вход получает строку row
# возвращает отношение значения в первом столбце к сумме значений в первом и во втором столбцах, умноженное на 100
# полученное отношение округляется до двух знаков после запятой
# возможное деление на ноль обработаем конструкцией try ... except ... 
# в случае деления на ноль функция вернет строку 'деление на ноль'
def return_rate(row):
   try:
        return round( ( row[0] / ( row[0] + row[1] ) ) * 100 , 2)
   except:
        return 'деление на ноль'

# определим функцию total_group, которая на вход получает строку row
# возвращает сумму значений в первом и во втором столбцах    
def total_group(row):
    return  row[0] + row[1] 

# в переменную total запишем общее количество строк в датасете df
total = df.shape[0]

### Проверка гипотезы №1

<div class="alert alert-block alert-info">
<font color='Black'>

Проверим гипотезу №1. Количество детей клиента влияет на факт погашения кредита в срок.

Сформируем сводную таблицу по столбцу `children`, будем группировать по значениям столбца `debt` - целевому признаку, в качестве значений для сводной таблицы в нашем случае можно использовать значения любого столбца, будем использовать значения столбца `gender` и считать количество его значений. Далее в результирующую таблицу добавим столбцы:
- `return_rate_%` - процент возврата кредитов в срок по группе,
- `group_%` - доля группы во всей выборке.

In [48]:
# строим сводную таблицу по исходному датафрейму df по столбцу `children`, 
# группируем по значениям столбца `debt` 
# и считаем значения в столбце `gender`, используя метод count()
df_pivot_children = df.pivot_table(index=['children'], columns='debt', values='gender', aggfunc='count')

# заполняем пропуски значением 0, чтобы результирующие столбцы рассчитывались корректно
df_pivot_children = df_pivot_children.fillna(0)

# добавляем в сводную таблицу столбцы return_rate_% и group_%, используя функции 
# return_rate, которая возвращает процент возврата кредитов в срок по группе 
# и total_group, которая возвращает общий размер выборки по группе
df_pivot_children['return_rate_%'] = df_pivot_children.apply(return_rate, axis=1)
df_pivot_children['group_%'] = df_pivot_children.apply(total_group, axis=1)

# преобразуем значение в столбце group_% в долую группе во всей выборке, используя размер всей выборке
# умножим полученное значение на 100 и округлим до двух знаков после запятой с помощью round()
df_pivot_children['group_%'] = round(df_pivot_children['group_%'] / total * 100, 2)

# выведем итоговую сводную таблицу, отсортировав по убыванию значений в return_rate_%
df_pivot_children.sort_values(by='return_rate_%', ascending=False)

debt            0       1  return_rate_%  group_%
children                                         
5             9.0     0.0         100.00     0.04
0         13112.0  1071.0          92.45    66.06
3           303.0    27.0          91.82     1.54
1          4411.0   445.0          90.84    22.62
2          1858.0   194.0          90.55     9.56
4            37.0     4.0          90.24     0.19

<div class="alert alert-block alert-info">
<font color='Black'>

Мы видим, что на первом месте по проценту возврата кредитов в срок находитcя группа клиентов с 5-ю детьми, процент возврата по данной группе составил 100%. 
При этом отметим, что доля группы клиентов с 5-ю детьми в выборке составляет всего 0,04%, доля данной группы не существенна для всей выборки в целом. Тоже утверждение справедливо и для группы клиентов с 4-мя детьми и 3-мя детьми. Поэтому для целей исследования не будем рассматривать результаты по данным группам.

В отношении групп клиентов без детей, с 1-м ребенком или 2-мя детьми можно сделать вывод, что процент возврата кредитов в срок при отсутствии детей составляет 92,45%, при наличии одного ребенка или двух детей процент возврата кредитов в срок примерно одинаковый, по группе с одним ребенком чуть выше 90,84%, по группе с двумя детьми 90,55%.

<div style="border:solid green 3px; padding: 20px">
    
Можно сделать вывод, что количество детей влияет на возврат кредита в срок, отсутствие детей повышает вероятность возврата кредита в срок, наличие детей понижает данную вероятность. Но данная зависимость неярко выражена, отличие в проценте возврата кредита в срок между группами составляет не более 2%. Кроме того, группы клиентов с более чем двумя детьми оказались не репрезентативными и поэтому результаты по данным группам не принимались во внимание при формировании вывода.
</div>   

### Проверка гипотезы №2

<div class="alert alert-block alert-info">
<font color='Black'>

2. Семейное положение клиента влияет на факт погашения кредита в срок.

Сформируем сводную таблицу по столбцу `family_status_id`, будем группировать по значениям столбца `debt` - целевому признаку, в качестве значений для сводной таблицы в нашем случае можно использовать значения любого столбца, будем использовать значения столбца `gender` и считать количество его значений. Далее в результирующую таблицу добавим столбцы:
- `family_status` - семейное положение,
- `return_rate_%` - процент возврата кредитов в срок по группе,
- `group_%` - доля группы во всей выборке.


In [49]:
# строим сводную таблицу по исходному датафрейму df по столбцу `family_status_id`, 
# группируем по значениям столбца `debt` 
# и считаем значения в столбце `gender`, используя метод count()
df_pivot_family_id = df.pivot_table(index=['family_status_id'],columns='debt', values='gender', aggfunc='count')

# добавляем в сводную таблицу столбцы return_rate_% и group_%, используя функции 
# return_rate, которая возвращает процент возврата кредитов в срок по группе 
# и total_group, которая возвращает общий размер выборки по группе
df_pivot_family_id['return_rate_%'] = df_pivot_family_id.apply(return_rate, axis=1)
df_pivot_family_id['group_%'] = df_pivot_family_id.apply(total_group, axis=1)

# преобразуем значение в столбце group_% в долую группе во всей выборке, используя размер всей выборке
# умножим полученное значение на 100 и округлим до двух знаков после запятой с помощью round()
df_pivot_family_id['group_%'] = round(df_pivot_family_id['group_%'] / total * 100, 2)

# дополним сводную таблицу данными о семейном положении
# воспользуемся "словарем" family_status_dict и по family_status_id добавим family_status
# используем для присоединения family_status_dict к df_pivot_family_id метод merge()
# результат сохраним в df_pivot_family_status
df_pivot_family_status = df_pivot_family_id.merge(family_status_dict, on='family_status_id', how='left')

# переименуем столбцы в df_pivot_family_status для наглядности результата
df_pivot_family_status = df_pivot_family_status.rename(columns={0:'debt_0', 1:'debt_1'})

# выведем итоговую сводную таблицу, отсортировав по убыванию значений в return_rate_%
df_pivot_family_status.sort_values(by='return_rate_%', ascending=False)

family_status_id  debt_0  debt_1  return_rate_%  group_%  \
2                 2     896      63          93.43     4.47   
3                 3    1110      85          92.89     5.57   
0                 0   11413     931          92.46    57.49   
1                 1    3775     388          90.68    19.39   
4                 4    2536     274          90.25    13.09   

           family_status  
2         вдовец / вдова  
3              в разводе  
0        женат / замужем  
1       гражданский брак  
4  Не женат / не замужем

<div class="alert alert-block alert-info">
<font color='Black'>

На первом месте по проценту возврата кредитов в срок находитcя группа клиентов со стасум вдовец/вдова, процент возврата по данной группе составил 93,43%. На втором и третьем местах с небольшим отставанием находятся группы со статусом в разводе и женат/замужем, процент возврата кредитов в срок по данным группам 92,89% и 92,46% соответственно.
При этом отметим, что доля группы клиентов со статусами вдовец/вдова и в разводе в выборке составляет около 5%, то есть данные группы в выборке малочисленны.

В отношении групп клиентов со статусами гражданский брак и не женат/не замужем процент возврата кредитов в срок примерно на одном уровне, 90,68% и 90,25% соответственно.

<div style="border:solid green 3px; padding: 20px">
    
Семейный статус влияет на возврат кредита в срок, факт того, что клиент, хотя бы один раз состоял в официально зарегистрированном браке либо состоит в нем на момент рассмотрения заявки на кредит, повышает вероятность возврата кредита в срок, отсутствие официально зарегистрированных отношений снижает вероятность возврата кредита в срок. Но данная зависимость неярко выражена, отличие в проценте возврата кредита в срок между группами составляет не более 3,2%.
</div> 

### Проверка гипотезы №3

<div class="alert alert-block alert-info">
<font color='Black'>

3. Уровень дохода клиента влияет на факт погашения кредита в срок.

Сформируем сводную таблицу по столбцу `total_income_category`, будем группировать по значениям столбца `debt` - целевому признаку, в качестве значений для сводной таблицы в нашем случае можно использовать значения любого столбца, будем использовать значения столбца `gender` и считать количество его значений. Далее в результирующую таблицу добавим столбцы:
- `return_rate_%` - процент возврата кредитов в срок по группе,
- `group_%` - доля группы во всей выборке.

In [50]:
# строим сводную таблицу по исходному датафрейму df по столбцу `total_income_category`, 
# группируем по значениям столбца `debt` 
# и считаем значения в столбце `gender`, используя метод count()
df_pivot_income = df.pivot_table(index=['total_income_category'],columns='debt', values='gender', aggfunc='count')

# добавляем в сводную таблицу столбцы return_rate_% и group_%, используя функции 
# return_rate, которая возвращает процент возврата кредитов в срок по группе 
# и total_group, которая возвращает общий размер выборки по группе
df_pivot_income['return_rate_%'] = df_pivot_income.apply(return_rate, axis=1)
df_pivot_income['group_%'] = df_pivot_income.apply(total_group, axis=1)

# преобразуем значение в столбце group_% в долую группе во всей выборке, используя размер всей выборке
# умножим полученное значение на 100 и округлим до двух знаков после запятой с помощью round()
df_pivot_income['group_%'] = round(df_pivot_income['group_%'] / total * 100, 2)

# выведем итоговую сводную таблицу, отсортировав по убыванию значений в return_rate_%
df_pivot_income.sort_values(by='return_rate_%', ascending=False)

debt                       0     1  return_rate_%  group_%
total_income_category                                     
D                        329    21          94.00     1.63
B                       4685   356          92.94    23.48
A                         23     2          92.00     0.12
C                      14673  1360          91.52    74.67
E                         20     2          90.91     0.10

<div class="alert alert-block alert-info">
<font color='Black'>

Обратим внимание на то, что группы клиентов с категориями дохода E,D,A составляют в общей выборке не более 1,63% каждая, данные группы не репрезентативны, поэтому мы не можем рассматривать их с целью формирования выводов.
В отношении групп клиентов со категориями доходов C от 50001 до 200000 и B от 200001 до 1000000 процент возврата кредита в срок составляет 91,52% и 92,94% соответственно. 

<div style="border:solid green 3px; padding: 20px">
Уровень дохода клиента влияет на возврат кредита в срок, чем выше доход, тем выше вероятность возврата кредита в срок. Но данная зависимость неярко выражена, отличие в проценте возврата кредита в срок между группами составляет не более 1,5%. При этом в выборке имеются не репрезентативные группы с категориями E от 0 до 30000, D от 30001 до 50000, A от 1000001 и выше, поэтому результаты исследования по данным группам не принимались во внимание при формировании вывода.
</div>   

### Проверка гипотезы №4

<div class="alert alert-block alert-info">
<font color='Black'>

4. Цель кредита влияет на факт погашения кредита в срок.

Сформируем сводную таблицу по столбцу `purpose_category`, будем группировать по значениям столбца `debt` - целевому признаку, в качестве значений для сводной таблицы в нашем случае можно использовать значения любого столбца, будем использовать значения столбца `gender` и считать количество его значений. Далее в результирующую таблицу добавим столбцы:
- `return_rate_%` - процент возврата кредитов в срок по группе,
- `group_%` - доля группы во всей выборке.

In [51]:
# строим сводную таблицу по исходному датафрейму df по столбцу `purpose_category`, 
# группируем по значениям столбца `debt` 
# и считаем значения в столбце `gender`, используя метод count()
df_pivot_purpose = df.pivot_table(index=['purpose_category'],columns='debt', values='gender', aggfunc='count')

# добавляем в сводную таблицу столбцы return_rate_% и group_%, используя функции 
# return_rate, которая возвращает процент возврата кредитов в срок по группе 
# и total_group, которая возвращает общий размер выборки по группе
df_pivot_purpose['return_rate_%'] = df_pivot_purpose.apply(return_rate, axis=1)
df_pivot_purpose['group_%'] = df_pivot_purpose.apply(total_group, axis=1)

# преобразуем значение в столбце group_% в долую группе во всей выборке, используя размер всей выборке
# умножим полученное значение на 100 и округлим до двух знаков после запятой с помощью round()
df_pivot_purpose['group_%'] = round(df_pivot_purpose['group_%'] / total * 100, 2)

# выведем итоговую сводную таблицу, отсортировав по убыванию значений в return_rate_%
df_pivot_purpose.sort_values(by='return_rate_%', ascending=False)

debt                          0    1  return_rate_%  group_%
purpose_category                                            
операции с недвижимостью  10032  782          92.77    50.37
проведение свадьбы         2149  186          92.03    10.88
получение образования      3644  370          90.78    18.69
операции с автомобилем     3905  403          90.65    20.06

<div class="alert alert-block alert-info">
<font color='Black'>

На первом месте по проценту возврата кредита в срок находятся операции с недвижимостью 92,77%, на втором - проведение свадьбы 92,03%, на третьем - получение образования 90,78%, на четвертом 90,65%.

<div style="border:solid green 3px; padding: 20px">
Цель кредита влияет на возврат кредита в срок, наиболее надежными являются заемщики, получающие кредит на операции с недвижимостью и проведение свадьбы, менее надежными являются заемщики, получающие кредит на операции с автомобилем и получение образования. Но данная зависимость неярко выражена, разница в проценте возврата кредита в срок между группами составляет не более 2,1%.
</div>   

## Общий вывод

<div style="border:solid green 3px; padding: 20px">

- **Краткий обзор проведённой работы** 
    
    Проанализировали данные статистики о платёжеспособности клиентов банка. Выполнили предобработку исходного датасета, обработали пропущенные и некорректные значения в данных, удалили дубликаты, провели категоризацию некоторых переменных. Рассчитали процент возврата кредита в срок по требуемым группам и выяснили, влияет ли количество детей, семейное положение, уровень дохода и цель кредита клиента на факт погашения кредита в срок.
    

- **Главные выводы** 
    
    1. Количество детей влияет на возврат кредита в срок, отсутствие детей повышает вероятность возврата кредита в срок, наличие детей понижает данную вероятность. Но данная зависимость неярко выражена, отличие в проценте возврата кредита в срок между группами составляет не более 2%. Кроме того, группы клиентов с более, чем двумя детьми, являются не репрезентативными и результаты по данным группам не учитывались при формировании вывода.
    
    2. Семейный статус влияет на возврат кредита в срок, факт того, что клиент, хотя бы один раз состоял в официально зарегистрированном браке или состоит в нем на момент рассмотрения заявки на кредит, повышает вероятность возврата кредита в срок, отсутствие официально зарегистрированных отношений снижает вероятность возврата кредита в срок. Но данная зависимость неярко выражена, отличие в проценте возврата кредита в срок между группами составляет не более 3,2%.
   
    3. Уровень дохода клиента влияет на возврат кредита в срок, чем выше доход, тем выше вероятность возврата кредита в срок. Но данная зависимость неярко выражена, отличие в проценте возврата кредита в срок между группами составляет не более 1,5%. При этом в выборке имеются не репрезентативные группы с категориями E от 0 до 30000, D от 30001 до 50000, A от 1000001 и выше , поэтому результаты исследования по данным группам не принимались во внимание при формировании вывода.
   
    4. Цель кредита влияет на возврат кредита в срок, наиболее надежными являются заемщики, получающие кредит на операции с недвижимостью и проведение свадьбы, менее надежными являются заемщики, получающие кредит на операции с автомобилем и получение образования. Но данная зависимость неярко выражена, разница в проценте возврата кредита в срок между группами составляет не более 2,1%.  
    

- **Рекомендации**
    
    При построении модели кредитного скоринга рекомендуется учитывать все проанализированные признаки: количество детей, семейное положение, уровень дохода и цель кредита. Следует дополнительно оценить степень влияния данных признаков на способность заемщика возвратить кредит в срок, возможно, какие-то признаки оказывают более сильное влияние на целевой признак. Например, от уровня дохода напрямую зависит возможность заемщика осуществлять платежи по кредиту в срок, от количества детей и их возраста зависит показатель долговой нагрузки заемщика. 

    
    
  
</div>   